### ISSUE TITLE
Nested fields objects are mutating

### ISSUE THREAD

**Issue by iNikNik** [erikras/redux-form#375]
Hello there! I found some little bug - nested fileds are mutating when you change something:

```javascript

class MyForm extends PureComponent {
    // ... some stuff
    render() {
        const { fileds: { someField, mainProduct } } = this.props;
        return (
            <div>
                <input {...someField} />
                <MainProductInput fields={mainProduct} />
            </div> 
        );
    }
}

class MainProductInput extends PureComponent {
    // ... some stuff
    render() { 
        // NEVER rerenders because fields objects always the same
        const { fileds: { title, other } } = this.props;
        return (
            <div>
                <input {...title} />
                <input {...other } />
            </div> 
        );
    }
}

reduxForm({
  form: 'mySuperForm',
  fields: [ 'someField', 'mainProduct.title',  'mainProduct.other' ],
})(MyForm)
```

I mean if you pass down whole nested field object - your children component will never rerenders.
I believe we should map `myFields` values and create a new objects for all nested fields here ->
https://github.com/erikras/redux-form/blob/master/src/readFields.js#L25

```javascript
function cloneFields(fields) {
   // mapValues like lodash.mapValues
  return mapValues(fields, field => isNested(field) ? cloneFields(field) : field);
}
const fieldObjects = cloneFields(myFields);
```

**PS guys, I want to tell you really big thanks! You're doing nice job with redux-form)**

**Comment by iNikNik** [comment]
If I'm doing like this:
```javascript
render() {
        const { fileds: { someField, mainProduct } } = this.props;
        return (
            <div>
                <input {...someField} />
                <MainProductInput fields={ {...mainProduct} } />
            </div> 
        );
    }
```
Everything works perfect.

**Comment by erikras** [comment]
Hmm... It is supposed to be creating new instances for any part of the subtree that changes.

Will investigate... :mag:

**Comment by aanchalgera** [comment]
We are facing the exact issue

**Comment by erikras** [comment]
Is this still an issue in v4?

**Comment by iNikNik** [comment]
@erikras I've just checked it - seem issue still here

**Comment by luisrudge** [comment]
@erikras anything I can do to help you figure this one out?

**Comment by kristian-puccio** [comment]
Just found this as well ```
	componentWillReceiveProps(newProps) {
		const isEqual = newProps.fields.customer.email === this.props.fields.customer.email;
		console.log(isEqual, "....isEqual....");
	}
```

I'll try my hand at a pull-request now and see how I go :)

**Comment by erikras** [comment]
^^ Look at that test. (it passes) What's wrong about it?

**Comment by kristian-puccio** [comment]
When I get isEqual = true and I've changed a field that's bad. It means that the field has mutated rather then create a new pointer that I can do a quick reference check against.

One thing I am noticing is that if I change the field so it isn't nested ie just email then I get `false "....isEqual...."`, which is correct.

**Comment by erikras** [comment]
Okay, but you looked at the test, right? The field that isn't changed is still `===`, but the one that does change is `!==`.

**Comment by iNikNik** [comment]
@erikras try to add test like [this](https://github.com/erikras/redux-form/commit/f0febaf8a2f8a64f5dd22940037bc8932fcce857#diff-61a744355e770a4a1c29bf259c518399R1970):
```javascript
const address = stub.props.fields.address
expect(stub.props.fields.address).toNotBe(address)
```

**Comment by erikras** [comment]
@iNikNik Very good. That's a bug for sure. Perhaps that's what was originally being referenced. :+1:

**Comment by luisrudge** [comment]
do you think fixing this will help with #529?

**Comment by erikras** [comment]
@luisrudge Probably not, since this mutating bug is _preventing_ rerendering. Fixing this will result in _more_ rendering.

**Comment by luisrudge** [comment]
Damn :D

**Comment by kristian-puccio** [comment]
But it will allow us to wrap components in purerendermixin so only the components that change will be re-rendered.

**Comment by kristian-puccio** [comment]
Oh and sorry I didn't look at the test. Felt a bit flakey last night so went to bed early. Hopefully tonight I'll get some time.

**Comment by erikras** [comment]
Fix published as [`v4.1.15`](https://github.com/erikras/redux-form/releases/tag/v4.1.15). Please confirm.

**Comment by kristian-puccio** [comment]
Perfect!

Thanks mate you are awesome

**Comment by bbenezech** [comment]
@erikras Sorry to bring bad news... :(

```javascript
  shouldComponentUpdate(nextProps) {
     console.log(this.props.prescription.comment.value === nextProps.prescription.comment.value); // always true
     console.log(this.props.prescription === nextProps.prescription); // always true
  }
```
When I update comment, both `this.props.prescription.comment.value` and `nextProps.prescription.comment.value` are equal in value and object (value is correct), which would mean that this.props.prescription has been mutated into nextProps.prescription.

I use `this.props.fields.prescriptions.addField(prescription);`, prescriptions is an array at the root of my reduxForm.

I use the latest release.

**Comment by bbenezech** [comment]
Ok, I wrote a spec for the first issue. @erikras Can you have a look?
And looks like there also is a mutation on the field array itself.

```javascript
  it('should not mutate tree or change untouched nested fields in arrays', () => {
    const store = makeStore();
    const form = 'testForm';
    const Decorated = reduxForm({
      form,
      fields: ['phones[]']
    })(Form);
    const dom = TestUtils.renderIntoDocument(
      <Provider store={store}>
        <Decorated/>
      </Provider>
    );
    const stub = TestUtils.findRenderedComponentWithType(dom, Form);

    stub.props.fields.phones.addField();
    stub.props.fields.phones.addField();

    const phones = stub.props.fields.phones;
    const phone0 = phones[0];
    const phone1 = phones[1];

    phone0.onChange('555-1234');

    expect(stub.props.fields.phones).toNotBe(phones); // fails. tree mutation?
    expect(stub.props.fields.phones[0]).toNotBe(phone0); // works.
    expect(stub.props.fields.phones[1]).toBe(phone1); // fails. useless update?
  });
```

**Comment by Coobaha** [comment]
@erikras Hi!

Right now you mutate object primitives, for `phones[].number` and `phone[0].number.onChange()`.
Phone array and all its members have same references (same objects),and  only field primitive props are mutated.

So this evaluates to true: `nextProps.fields === this.props.fields && nextProps.fields.phones === this.props.fields.phones && nextProps.fields.phones[0] === this.props.fields.phones[0]`.

Should fields and values be new objects/arrays on their part change, like you will do in the reducer on state change?

```
fields: {
...fields,
phones: [
    {
    ...phones[0],
    value: newValue,
    touched: true,
    (etc...)
    },
    ...without(phones, phone[0])
]},
```

This will allow pure rendering, and also passing whole deep form slices to components. Right now I am supposed to render whole fields in wrapped component.

Btw, props.values are changed on focus/blur events, should they? Imo they should be new object  for corresponding path only if value in this path changed, like fields above.

P.S. Sorry if this is already some where in issues, I did not found them.

**Comment by villanuevawill** [comment]
@erikras You're the man, and thanks for this library! Would you mind addressing the two questions above please?  Was this fixed in a later release? I'm still on version 4 and getting ready to update, but want to make sure if the SPEC @bbenezech would pass or @Coobaha 's question.  My app slows down because everything must re-render since the fields array shows it remains equal.

**Comment by erikras** [comment]
Published fix in [`v5.3.4`](https://github.com/erikras/redux-form/releases/tag/v5.3.4).
